sui_programmability/examples/capy/sources/capy.move

```diff
@@ -0,0 +1,408 @@
+        name: String,
+        values: vector<Value>
+    }
+
+    /// The Capy itself. Every Capy has its unique set of genes,
+    /// as well as generation and parents. Ownable, tradeable.
+    struct Capy has key, store {
+        id: UID,
+        gen: u64,
+        url: Url,
+        genes: Genes,
+        item_count: u8,
+        attributes: vector<Attribute>,
+    }
+
+    /// Belongs to the creator of the game. Has store, which
+    /// allows building something on top of it (ie shared object with
+    /// multi-access policy for managers).
+    struct CapyManagerCap has key, store { id: UID }
+
+    /// Every capybara is registered here. Acts as a source of randomness
+    /// as well as the storage for the main information about the gamestate.
+    struct CapyRegistry has key {
+        id: UID,
+        capy_born: u64,
+        capy_day: u64,
+        capy_hash: vector<u8>,
+        genes: vector<GeneDefinition>
+    }
+
+
+    // ======== Events =========
+
+    /// Event. When a new registry has been created.
+    /// Marks the start of the game.
+    struct RegistryCreated has copy, drop { id: ID }
+
+    /// Event. Emitted when a new Gene definition was added.
+    /// Helps announcing new features.
+    struct GeneDefinitionAdded has copy, drop {
+        name: String,
+        values: vector<Value>
+    }
+
+    /// Event. When new Capy is born.
+    struct CapyBorn has copy, drop {
+        id: ID,
+        gen: u64,
+        genes: Genes,
+        birthday: u64,
+        attributes: vector<Attribute>,
+        parent_one: ID,
+        parent_two: ID,
+        bred_by: address
+    }
+
+    /// Event. Emitted when a new item is added to a capy.
+    struct ItemAdded<phantom T> has copy, drop {
+        capy_id: ID,
+        item_id: ID
+    }
+
+    /// Event. Emitted when an item is taken off.
+    struct ItemRemoved<phantom T> has copy, drop {
+        capy_id: ID,
+        item_id: ID,
+    }
+
+
+    // ======== Functions =========
+
+    /// Create a shared CapyRegistry and give its creator the capability
+    /// to manage the game.
+    fun init(ctx: &mut TxContext) {
+        let id = object::new(ctx);
+        let capy_hash = hash(object::uid_to_bytes(&id));
+
+        emit(RegistryCreated { id: object::uid_to_inner(&id) });
+
+        transfer::transfer(CapyManagerCap { id: object::new(ctx) }, tx_context::sender(ctx));
+        transfer::share_object(CapyRegistry {
+            id,
+            capy_hash,
+            capy_born: 0,
+            capy_day: 0,
+            genes: vec::empty()
+        })
+    }
+
+
+    // ======= Admin Functions =======
+
+    /// This method is rather complicated.
+    /// To define a new set of attributes, Admin must send it in a format:
+    /// ```
+    /// name = b"name of the attribute"
+    /// definitions = [
+    ///     [selector_u8, ...name_bytes],
+    ///     [selector_u8, ...name_bytes]
+    /// ]
+    /// ```
+    entry fun add_gene(
```

Comment:
Out of curiosity, why would you opt for `entry fun` instead of `public entry fun`?